get users and assorted Okta changes

jeremyzitomer-usds opened the pull request:
## Related Issue or Background Info

Added mutation to get all users with role information

## Changes Proposed

- Added mutation
- Renamed/refactored some internal classes; notably, OrganizationRoles -> AuthorityBasedOrganizationRoles, and CurrentOrganizationRoles -> OrganizationRoles
- Updated User class
- Added new permission "MANAGE_USER_LIST"
- removed Okta integration from API tests
- added some demo/test Okta mocking

## Rationale for the OrganizationRoles renames
I realized that the way that the old "OrganizationRoles" class was really being used, was exclusively when grabbing an authority/claim from the JWT or Okta (what those external sources provide is always an org's external ID and the associated user roles). Hence the rationale to call these roles "authority-based". I recognize there's a bit of grey between authority and claims, if there's a better word for this, I'm all ears.
Also, the old "CurrentOrganizationRoles" class was identical to the object that was needed in order to return organization roles for users of other organizations (especially in the create/updateUser flows), and it became too limiting to require that such a class be used only to denote org roles for the caller's current org.

sharmaneil commented on backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/ApiUserResolver.java:
```diff
@@ -22,15 +28,35 @@
+
+	public List<User> getUsers() {
+		Organization org = _organizationService.getCurrentOrganization();
+		List<ApiUser> users = _organizationService.getUsersInCurrentOrg(OrganizationRole.USER);
```
Would this only get users with the role USER, or everyone? From a quick scan, the `OrganizationRole.USER` parameter is used to query the okta group -- are only users with role USER in that group?

I ask because i assume the graphql query for `users` would return all humans, not just USER humans.

jeremyzitomer-usds replied:
it gets everyone. everyone in an organization is in the USER Okta group for that org. what the lines after that do is assign the specific roles of all users based on the additional Okta groups (ADMIN, ENTRY_ONLY) they are in for that organization.

timothybest-usds commented on backend/build_and_push.sh:
```diff
@@ -13,5 +13,5 @@ echo "Tagged $ACR_TAG"
 # Log in to ACR
 # Uncomment for user login
-# az acr login --name simplereportacr
+az acr login --name simplereportacr
```
Is this change intentional?

jeremyzitomer-usds replied:
nope! thanks for catching :)

benwarfield-usds commented:
Thanks for the extra information on the renames—I agree that the swap is logical, though more than a little confusing on first read. 😁 

Based on your (correct!) description of what the original OrganizationRoles object does, I would call it `OrganizationRoleClaims` or `OrganizationClaims` depending on whether or not we think the same object will eventually house the facility-access information (I think probably yes, but it'll require a refactor if so).

I don't think using Authority in the name is helpful: it's only Authority-based as a matter of convenience based on the way Okta sets up their JWT integration, and once we move beyond the current (comparatively) simple role-management approach, we will have to move beyond (or possibly entirely abandon) the `getAuthorities` approach to querying the Authorization object.

benwarfield-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/OktaServiceConfig.java:
```diff
@@ -0,0 +1,46 @@
+            IdentityAttributes user = altUser.getIdentity();
+            String username = user.getUsername();
+            OrganizationRoleClaims orgRoles = 
+                    new OrganizationRoleClaims(setupProps.getOrganization().getExternalId(),
```
Couldn't this just be `demoUsers.getDefaultUser().getAuthorization().getOrganizationExternalId()`?

benwarfield-usds reviewed (commented):
I had some structural questions but we resolved those offline, and I think I've deleted all the comments related to them.

I had a lot of trouble getting a consecutive time block to do the whole thing and get my head around it, so my apologies for the slowness that caused—it's a lot of individual changes to navigate around, even if the straight line-count isn't as big as some.

I made a bunch of fairly small-scale comments that I am mostly just assuming you can figure out which of seem significant and which not—the major points I wanted to raise and which may not be obvious from what I said inline are:

* I'm confused about the way the demo-user configuration is being used to create fake-Okta, and how it's entangled with the OrganizationInitializationService: I think that's worth teasing apart in order to get to where we will need to get to on having a good local dev/demo setup. Obviously trying to get to a fully-fleshed-out multi-role demo is not in scope for this PR, but I think a little bit of work to keep those things from getting tangled up (especially since I just recently untangled them) is worthwhile—the suggestions I made around DemoUserConfiguration and EmptyOktaServiceImpl were aimed at that, but I'd be happy to discuss offline as well.
* I'm concerned that the tests that set up the user to be a superuser _and_ an org admin indicate that our permission model for those endpoints is broken, and I would like us to fix that ASAP. I'm not sure that's a blocker either, for this PR, but we need to prioritize that sometime sooner than "after we have nothing else to do."

jeremyzitomer-usds commented on backend/src/test/java/gov/cdc/usds/simplereport/service/BaseServiceTest.java:
```diff
@@ -108,8 +107,7 @@ protected void clearOktaUsers() {
             String groupName = g.getProfile().getName();
-            if (g.getType() == GroupType.OKTA_GROUP &&
-                        groupName.startsWith(_authorizationProperties.getRolePrefix())) {
+            if (groupName.startsWith(_authorizationProperties.getRolePrefix())) {
```
not sure how i removed it, my bad!

jeremyzitomer-usds replied:
(not that the entire code block will stick around much longer anyway)

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/api/model/User.java:
```diff
@@ -1,41 +1,35 @@
-        this.isAdmin = isAdmin;
-        Optional<OrganizationRole> effectiveRole = orgwrapper.flatMap(CurrentOrganizationRoles::getEffectiveRole);
+        this.roles = new ArrayList<>();
+        orgwrapper.map(OrganizationRoles::getGrantedRoles).map(roles::addAll);
```
i want to return the full role list. while it is currently true that the permissions for our roles are concentric circles around each other, that may not always be true, so returning multiple may be valuable and i'd rather not shoehorn us.

benwarfield-usds replied:
Makes sense. If we end up with overlapping circles we'll need to make sure we are careful about the assumption that USER is still the God-group.

jeremyzitomer-usds replied:
also, there may be some confusion here about orgwrapper being an Optional, not a List, but either way I'll clean it up to make Sonar less mad.

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/ApiUserResolver.java:
```diff
@@ -22,15 +28,35 @@
         return new User(currentUser, currentOrgRoles, isAdmin);
 	}
+
+	public List<User> getUsers() {
```
yeah i couldn't find a good way to push this down the stack without pushing the User api model object down the stack as well, which smelled bad. *shrug*

benwarfield-usds replied:
It may become more clear what's possible after other refactors are done and/or as other new code lands. No sense trying to move things around just for the sake of moving them, though.

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserPermission.java:
```diff
@@ -10,6 +10,7 @@
     EDIT_PATIENT,
     EDIT_FACILITY,
     EDIT_ORGANIZATION,
+    MANAGE_USER_LIST,
```
is there a reason to make it MANAGE_USERS over MANAGE_USER_LIST ? just for extra differentiation?

benwarfield-usds replied:
Well, it seems like this role will probably be empowered to edit the settings of individual users, not just the list of them. I'm not actually sure what managing the list means—managing roles, but not being able to edit users?

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:
```diff
@@ -105,6 +112,19 @@ public void assertFacilityNameAvailable(String testingFacilityName) {
+    @AuthorizationConfiguration.RequirePermissionManageUserList
+    public List<ApiUser> getUsersInCurrentOrg(OrganizationRole role) {
+        String organizationExternalId = getCurrentOrganization().getExternalId();
+        List<String> usernames = _oktaService.getAllUsernamesForOrganization(organizationExternalId, role);
```
will call the below method, will leave the second refactor point for my next PR since I've started that anyway

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:
```diff
@@ -86,12 +91,14 @@ public Organization getOrganization(String externalId) {
-        }
-        return Optional.ofNullable(getOrganization(orgExternalId));
+    @AuthorizationConfiguration.RequirePermissionManageUserList
+    public Optional<OrganizationRoles> getOrganizationRolesForUser(ApiUser apiUser) {
```
agreed, which is why i've already started them moving them over in my next PR 😄 will address this then.

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:
```diff
@@ -37,25 +39,28 @@
     private OrganizationRepository _repo;
     private FacilityRepository _facilityRepo;
     private ProviderRepository _providerRepo;
+    private ApiUserRepository _userRepo;
```
ditto re next PR.

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java:
```diff
@@ -123,25 +147,37 @@ public void deleteOrganization(String externalId) {
-    public String getOrganizationExternalIdForUser(String username) {
+    public Optional<OrganizationRoleClaims> getOrganizationRolesForUser(String username) {
+
         UserList users = _client.listUsers(username, null, null, null, null);
```
Okta 4 evr <3

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java:
```diff
@@ -123,25 +147,37 @@ public void deleteOrganization(String externalId) {
 
-        throw new IllegalGraphqlArgumentException("User is not in any Okta user groups");
+        if (orgExternalIds.size() != 1) {
+            LOG.warn("User is in {} Okta organizations, not 1", orgExternalIds.size());
```
eek, there's no notion of archiving groups in Okta. Either they live on (which feels not good, because then do the old user-group relationships remain? and then does the OktaService always return all groups--existing and archived--that the user belongs to?) or we get rid of them. I'd prefer the latter, which means this code stays as-is. let me know if you feel strongly.

benwarfield-usds replied:
I feel strongly that we shouldn't have two conflicting definitions of "valid" in different places in the app—I think there are reasonable justifications for the approach that I took originally (well, obviously I do), though I also see your point about cleaning up quickly—worst case this won't be relevant for a significant while, so let's talk about it next week.

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java:
```diff
@@ -84,7 +93,22 @@ public void createUser(IdentityAttributes userIdentity, String organizationExter
+        String groupName = generateGroupName(externalId, role);
+
+        GroupList groups = _client.listGroups(groupName, null, null);
+        if (!groups.iterator().hasNext()) {
```
i know, i confessed my sins after writing this, but they're not really mine. alas.

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java:
```diff
@@ -84,7 +93,22 @@ public void createUser(IdentityAttributes userIdentity, String organizationExter
                 .buildAndCreate(_client);
     }
 
+    public List<String> getAllUsernamesForOrganization(String externalId, OrganizationRole role) {
```
i don't love that this change passes a db model object into the OktaService class, but i think your upsides outweigh the downsides.

jeremyzitomer-usds commented on backend/src/test/java/gov/cdc/usds/simplereport/api/BaseApiTest.java:
```diff
@@ -182,6 +129,23 @@ protected ObjectNode runQuery(String queryFileName) {
+     * expected error (either none or a single specific error message), and return
+     * the {@code data} section of the response if the error was as expected.
+     */
+    protected ObjectNode runQuery(String queryFileName, String expectedError) {
```
is it okay to pass in `variables` as null?

benwarfield-usds replied:
I don't know either, actually. My guess is that it's fine but I'd want to check, clearly.

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceEmptyImpl.java:
```diff
@@ -1,36 +1,54 @@
+        this.usernameRolesMap = usernameRolesMap;
+        this.orgRoleUsernamesMap = new HashMap<>();
+        for (Map.Entry<String,OrganizationRoleClaims> entry : usernameRolesMap.entrySet()) {
+            Map<OrganizationRole,List<String>> roleUsernamesMap = 
+                    orgRoleUsernamesMap.getOrDefault(entry.getValue().getOrganizationExternalId(),
+                                                     new HashMap<>());
```
i am going to gently decline for a) readability and b) because i think this code block doesn't address the case of adding usernames to `roleUsernamesMap` when `roleUsernamesMap` is already non-null; and after we add that back in, what i have now feels a touch more straightforward. considering this is just a demo class, i'm going to leave as is. (though I'll pull out orgId as its own variable, to make things a touch more readable.)

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationInitializingService.java:
```diff
@@ -72,6 +80,14 @@ public void initAll() {
+				.map(DemoUserConfiguration.DemoAlternateUser::getIdentity).collect(Collectors.toList());
+		for (IdentityAttributes user : users) {
+			_apiUserRepo.save(new ApiUser(user.getUsername(), user));
+			initOktaUser(user, emptyOrg.getExternalId());
```
so this is what Dev is -- its profiles include create-sample-data but NOT no-okta-mgmt. We can decide to change that down the line if we'd like.

benwarfield-usds replied:
Ah, I see. I think that should probably not be the default for dev, but we can talk it through as we go.

jeremyzitomer-usds commented on backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:
```diff
@@ -83,69 +85,215 @@ void whoami_nobody_okResponses() {
-            .put("suffix", "III")
-            .put("email", USERNAMES.get(0))
-            .put("organizationExternalId", _initService.getDefaultOrganizationId());
+        useOrgAdmin();
```
this (and the other tests where both superuser and org admin are invoked) is an unintentional byproduct of the various security restrictions on the methods involved in creating/updating users, but it's something i'm addressing remedying in my next PR so i'm going to leave it for then

jeremyzitomer-usds commented on backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:
```diff
@@ -83,69 +85,215 @@ void whoami_nobody_okResponses() {
+
+    @Test
+    void getUsers_orgUser_failure() {
+        useSuperUser();
```
ditto

jeremyzitomer-usds commented on backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:
```diff
@@ -83,69 +85,215 @@ void whoami_nobody_okResponses() {
+        useSuperUser();
+        useOrgAdmin();
+
+        when(_oktaService.getOrganizationRolesForUser(USERNAMES.get(0)))
```
this will be fixable once i resolve the superuser/orgadmin dual dependency in the next PR.

jeremyzitomer-usds replied:
the perils of copypasta

jeremyzitomer-usds commented on backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:
```diff
@@ -83,69 +85,215 @@ void whoami_nobody_okResponses() {
     }
 
+    @Test
+    void updateUser_orgUser_failure() {
```
ditto

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceEmptyImpl.java:
```diff
@@ -1,36 +1,54 @@
+        return orgRoleUsernamesMap.get(externalId).get(role);
     }
 
+    public void createOrganization(String name, String externalId) {}
```
yours and mine both!

jeremyzitomer-usds commented:
@benwarfield-usds I've addressed your PR feedback. With the caveat that several of your suggestions are things I'm addressing in my next PR, please let me know if there are any outstanding concerns before I merge.

And--kindly ignore Sonar's complaints about code coverage. I think it's ignoring that the Okta live repository class was renamed, and doesn't like that that repository isn't unit tested (which, frankly, it should take up with Okta, not me). But including my tests in `ApiUserManagementTest.java`, I'm pretty darn sure relative code coverage for this PR is much higher than Sonar would have you think.

benwarfield-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/OrganizationRole.java:
```diff
@@ -70,4 +69,8 @@ public int compare(OrganizationRole one, OrganizationRole other) {
         }
     }
+
+    public static final OrganizationRole getDefault() {
```
👍

benwarfield-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/DemoOktaRepository.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    public void createOrganization(String name, String externalId) {
+        orgRoleUsernamesMap.putIfAbsent(externalId, new HashMap<>());
```
I... did not realize that method was in Map. 👍

benwarfield-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:
```diff
@@ -84,7 +100,22 @@ public void createUser(IdentityAttributes userIdentity, String organizationExter
+
+        GroupList groups = _client.listGroups(groupName, null, null);
+        if (!groups.iterator().hasNext()) {
+            LOG.warn("Okta group is nonexistent; returning zero usernames");
```
Maybe worth indicating which org and role the group is missing for?

benwarfield-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:
```diff
@@ -123,25 +154,37 @@ public void deleteOrganization(String externalId) {
-                    groupName.endsWith(generateRoleSuffix(OrganizationRole.USER))) {
-                return getOrganizationExternalIdFromGroupName(groupName, 
-                                                              OrganizationRole.USER);
+            for (OrganizationRole role : OrganizationRole.values()) {
```
I think this code path and the one in OrganizationExtractor should not both exist in the long term, but I think this is in keeping with the other question about longer-term business rules for inactive orgs, and also potentially something we're going to refactor when we do V2 access control anyway.

jeremyzitomer-usds replied:
Yeah, and this is designed specifically for users that are *not* the current user

benwarfield-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/DemoOktaRepository.java:
```diff
@@ -0,0 +1,91 @@
+import gov.cdc.usds.simplereport.service.model.IdentityAttributes;
+
+/**
+ * Created by jeremyzitomer-usds on 1/7/21
```
In the department of ridiculous nitpicks, I would turn off this auto-generation (or not type it in by hand)—it's all there in the git history anyway, and it's not really information that you need when using the public API of this class, which is what the javadoc should tell us.

jeremyzitomer-usds replied:
</3

benwarfield-usds commented on backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:
```diff
@@ -83,69 +89,191 @@ void whoami_nobody_okResponses() {
+    }
+
+    private OrganizationRoleClaims getOrgRoles(OrganizationRole role) {
+        Set<OrganizationRole> roles = new HashSet<>();
+        roles.add(OrganizationRole.USER);
+        roles.add(role);
+        return new OrganizationRoleClaims(_initService.getDefaultOrganization().getExternalId(),
+                                                   roles);
```
```suggestion
        return new OrganizationRoleClaims(_initService.getDefaultOrganization().getExternalId(),
                                                   Set.of(OrganizationRole.USER, role));
```

Or not, honestly, this is fine. 🙂

jeremyzitomer-usds replied:
needs to be able to de-dupe duplicates though! Set can't, but HashSet can.

benwarfield-usds replied:
Really? Set.of chokes on duplicates? I would not have guessed that! (Obviously, I suppose!)

benwarfield-usds commented on backend/src/test/java/gov/cdc/usds/simplereport/idp/repository/OktaRepositoryTest.java:
```diff
@@ -0,0 +1,61 @@
+import gov.cdc.usds.simplereport.config.authorization.OrganizationRoleClaims;
+import gov.cdc.usds.simplereport.service.model.IdentityAttributes;
+
+class OktaRepositoryTest {
```
I think it might be helpful to add a javadoc saying what these tests are and why we aren't running them.

I'd also love to get them running, at least optionally, but that's a separate question.

jeremyzitomer-usds replied:
looking to get them running in v2

benwarfield-usds reviewed (approved):
LGTM!

I think.

That's a lot of code—congrats (possibly premature) on getting it over the finish line!
